internal/startup_manager/startup_manager.go

```diff
@@ -32,21 +40,167 @@ func NewStartupManager() *StartupManager {
 	}
 }
 
+var previousCollector prometheus.Collector = nil
+
+func (manager *StartupManager) StartAppServer() {
+	config := manager.appConfig
+
+	host := config.Server.Host
+	port := strconv.Itoa(config.Server.Port)
+	appServer := server.NewAppServer(host, port, config, config.Logstash.HttpTimeout)
+
+	slog.Info("starting server on", "host", host, "port", port)
+	if err := appServer.ListenAndServe(); err != nil {
+		slog.Error("failed to listen and serve", "err", err)
+		os.Exit(1)
+	}
+}
+
+func (manager *StartupManager) SetupHotReload(ctx context.Context) error {
+	var (
+		runGroup      run.Group
+		reloadManager = reload.NewManager()
+	)
+
+	// Add all app reloaders in order.
+	reloadManager.Add(0, reload.ReloaderFunc(func(ctx context.Context, id string) error {
+		// If configuration fails ignore reload with a warning.
+		slog.Info("ev", "event", id)
+		if (id == "no-event") {
+			return nil
+		}
+
+		err := manager.LoadConfig(ctx)
+		if err != nil {
+			slog.Warn("Config could not be reloaded", "err", err)
+			return err
+		}
+
+		slog.Info("Config reloaded")
+		return nil
+	}))
+
+	reloadManager.Add(100, reload.ReloaderFunc(func(ctx context.Context, id string) error {
+		if (id == "no-event") {
+			return nil
+		}
+
+		manager.SetupPrometheus(ctx)
+		slog.Info("Prometeus reloaded")
+		return nil
+	}))
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	runGroup.Add(
+		func() error {
+			slog.Info("Starting reload manager")
+			return reloadManager.Run(ctx)
+		},
+		func(_ error) {
+			slog.Info("Stopping reload manager")
+			cancel()
+		},
+	)
+
+
+	config := manager.appConfig
+
+	host := config.Server.Host
+	port := strconv.Itoa(config.Server.Port)
+	appServer := server.NewAppServer(host, port, config, config.Logstash.HttpTimeout)
+
+	runGroup.Add(
+		func() error {
+			slog.Info("starting server on", "host", host, "port", port)
+			return appServer.ListenAndServe()
+		},
+		func(_ error) {
+			slog.Info("Stopping HTTP server")
+			ctx, cancel := context.WithTimeout(context.Background(), config.Logstash.HttpTimeout)
+			defer cancel()
+			err := appServer.Shutdown(ctx)
+			if err != nil {
+				slog.Error("Could not shut down http server", "err", err)
+			}
+			os.Exit(1)
+		},
+	)
+
+
+	// File watcher:
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	err = watcher.Add(filepath.Dir(*manager.flagsConfig.configLocation))
+	fname := filepath.Base(*manager.flagsConfig.configLocation)
+	if err != nil {
+		slog.Warn("could not add file watcher for %s: %s", *manager.flagsConfig.configLocation, err)
+		return err
+	}
+
+	// Add file watcher based reload notifier.
+	reloadManager.On(reload.NotifierFunc(func(ctx context.Context) (string, error) {
```

Comment:
if the file gets written, but no config gets changed this still reloads the server, right? it may not be the best approach

Reply:
yes
but you should parse it first
so
- file was written with the same content- ignore it
- file was written with different content, but the struct with config was not changed (eg. order was changed, whitespace was added)
you should ignore that